packages/microcosm/src/subject.js

```diff
@@ -158,87 +137,151 @@ export class Subject {
+      let value = obj[key]
 
-    return this.subscribe({
-      next: dispatch,
-      error: dispatch,
-      complete: dispatch,
-      cancel: dispatch
-    })
-  }
+      if (!isObject(value)) {
+        payload = set(payload, key, value)
+        continue
+      }
 
-  static hash(obj: *): Subject {
-    if (obj instanceof Subject) {
-      return obj
-    }
+      jobs += 1
 
-    let subject = new Subject()
+      if (value instanceof Subject) {
+        payload = set(payload, key, value.payload)
+      }
 
-    if (getObservable(obj)) {
-      obj.subscribe(subject)
-      return subject
-    }
+      let subscription = Subject.hash(value).subscribe({
+        next: assign.bind(null, key),
+        error: subject.error,
+        complete: taskFinished,
+        cancel: taskFinished
+      })
 
-    if (obj == null || typeof obj !== 'object') {
-      subject.next(obj)
-      subject.complete()
-      return subject
+      subject.subscribe({ cancel: subscription.unsubscribe })
     }
 
-    if (typeof obj.then === 'function') {
-      subject.subscribe(Observable.fromPromise(obj).subscribe(subject))
-      return subject
+    if (payload !== subject.payload) {
+      subject.next(payload)
     }
 
-    let keys = Object.keys(obj)
-    let payload = Array.isArray(obj) ? [] : {}
-    let jobs = keys.length
+    building = false
 
-    subject.next(payload)
+    if (jobs <= 0) {
+      subject.complete()
+    }
 
-    function complete() {
-      if (--jobs <= 0) {
-        subject.complete()
-      }
+    return subject
+  }
+
+  static from(source: *): Subject {
+    if (source instanceof Subject) {
+      return source
     }
 
-    function assign(key: string, value: *): void {
-      // $FlowFixMe - Signature of set() is too strict
-      payload = set(payload, key, value)
+    if (isObservable(source)) {
+      return fromObservable(source)
+    }
 
-      if (payload !== subject.payload) {
-        subject.next(payload)
-      }
+    if (isPromise(source)) {
+      return fromPromise(source)
     }
 
-    for (var i = 0, len = keys.length; i < len; i++) {
-      let key = keys[i]
+    return Subject.of(source)
+  }
 
-      let subscription = Observable.wrap(obj[key]).subscribe({
-        next: value => assign(key, value),
-        complete: complete,
-        error: subject.error
-      })
+  // $FlowFixMe This function enumerates on variable arguments
+  static of() {
+    let subject = new Subject()
 
-      subject.subscribe(subscription)
+    for (var i = 0; i < arguments.length; i++) {
+      subject.next(arguments[i])
     }
 
+    subject.complete()
+
     return subject
   }
 }
+
+function fromPromise(promise: Promise<*>): Subject {
+  let subject = new Subject()
+
+  promise
+    .then(subject.next)
+    .then(() => subject.complete())
+    .catch(subject.error)
+
+  return subject
+}
+
+function fromObservable(observable: Observable): Subject {
+  let subject = new Subject()
+  observable.subscribe(subject)
```

Comment:
does this mutate the `subject` in place?

Reply:
It places the subject downstream to the observable. Because Subjects adhere to the subscriber interface, they can  be passed straight into subscribe. This is the same as:

```javascript
observable.subscribe({
  next: value => subject.next(value),
  error: reason => subject.error(reason),
  complete: () => subject.complete()
  cancel: () => subject.cancel()
})
```

Reply:
I should make an interface for this actually. Both Subscription and Subject should implement it. Jotting that down...